### PR TITLE
User Account Control hardening

### PR DESCRIPTION
**Pull request by obsti8383**
Sets User Account Control (UAC) for users with administrative rights to the highest setting. This prevents a mechanism used by attackers to circumvent UAC (see for instance: http://www.securityweek.com/windows-uac-bypassed-using-event-viewer, https://www.heise.de/security/meldung/Erpressungs-Trojaner-Erebus-umgeht-erfolgreich-UAC-Abfrage-von-Windows-3619820.html).

Doesn't automatically merge because I had to adapt the percentages for progress.SetValue().

### CODE REVIEW

**Comment by obsti8383** [comment]
- Merged with master
- Migrated build script to go1.8 stable by verifing version via "go version" and just calling "go" as a binary (so it works with current stable linux 1.8 build"

File: uac.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+    "golang.org/x/sys/windows/registry"
+)
+
+func trigger_uac(enable bool) {
+    key, _ := registry.OpenKey(registry.LOCAL_MACHINE, "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Policies\\System", registry.WRITE)
+
+    if enable {
+        events.AppendText("Setting UAC to not ask for changing Windows settings (default)\n")
```
Comment (ccoenen) [inline uac.go]: could it be that the labels have been swapped? this reads to me as if "enabled" branch would set it to default (where simple settings change does not trigger UAC) - not sure if it's only the message or also the registry key.

File: uac.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        err := key.SetDWordValue("ConsentPromptBehaviorAdmin", 5)
+        if err != nil {
+            events.AppendText("!! Setting UAC failed.\n")
+        }
+    } else {
+        events.AppendText("Setting UAC to always ask on secure desktop (secure)\n")
```
Comment (ccoenen) [inline uac.go]: This reads like it would set it to more secure in "else" (disable).

**Comment by ccoenen** [comment]
the comments above are just related to the lines directly above it (29 and 35 in those cases). It might just be a language barrier problem for me, please don't be offended, I might well be wrong.

File: uac.go (hunk 2 above)
Comment (obsti8383) [inline uac.go]: Hi @ccoenen: Thats a little bit disturbing, yes. If you check main.go you can see that calling the trigger functions with enable=false this actually means "do the hardening" and if you use "true" that means restore the original settings ("enable" the feature that has been disabled due to hardening).

Comment (ccoenen) [inline uac.go]: That is so backwards That I would consider it a bug. An API should not be surprising. enable=false _must_ mean disable. So either eht main.go is wrong or the naming of the parameter. Should I open another ticket for it?

**Comment by botherder** [comment]
This pull request highlights that perhaps we need to restructure a little bit the "logic" of the code. The choice of words might have been not great. Shall I just merge this and then we do a little restructuring?

**Comment by obsti8383** [comment]
Yes we should, thanks!

**Pull request merged**